Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,564 @@
+
+    case let .active(_, _, statusPromise):
+      statusPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+    case .completed:
+      self.interceptors = nil
+    }
+  }
+
+  // MARK: - Interceptors to User Function
+
+  @inlinable
+  internal func receiveInterceptedPart(_ part: GRPCServerRequestPart<Request>) {
+    switch part {
+    case let .metadata(headers):
+      self.receiveInterceptedMetadata(headers)
+    case let .message(message):
+      self.receiveInterceptedMessage(message)
+    case .end:
+      self.receiveInterceptedEnd()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      let statusPromise: EventLoopPromise<GRPCStatus> = context.eventLoop.makePromise()
+
+      // Make a context to invoke the user handler with.
+      let context = GRPCAsyncServerCallContext(
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef
+      )
+
+      // Create a request stream to pass to the user function and capture the
+      // handler in the updated state to allow us to produce more results.
+      let requestStream = GRPCAsyncRequestStream<Request>(AsyncThrowingStream { continuation in
+        self.state = .active(
+          { streamEvent in
+            switch streamEvent {
+            case let .message(request): continuation.yield(request)
+            case .end: continuation.finish()
+            }
+          },
+          context,
+          statusPromise
+        )
+      })
+
+      // Create a writer that the user function can use to pass back responses.
+      let responseStreamWriter = GRPCAsyncResponseStreamWriter(
+        context: context,
+        compressionIsEnabled: self.context.encoding.isEnabled,
+        send: self.interceptResponse(_:metadata:)
+      )
+
+      // Send response headers back via the interceptors.
+      self.interceptors.send(.metadata([:]), promise: nil)
+
+      // Register callbacks on the status future.
+      statusPromise.futureResult.whenComplete(self.userFunctionStatusResolved(_:))
+
+      // Spin up a task to call the async user handler.
+      self.task = statusPromise.completeWithTask {
+        // Check for cancellation before calling the user function.
+        // This could be the case if the RPC has been cancelled or had an error before this task
+        // has been scheduled.
+        guard !Task.isCancelled else {
+          throw CancellationError()
+        }
+
+        // Call the user function.
+        try await self.userHandler(requestStream, responseStreamWriter, context)
+
+        // Check for cancellation after the user function has returned so we don't return OK in the
+        // event the RPC was cancelled. This is probably overkill because the `handleError(_:)`
+        // will also fail the `statusPromise`.
+        guard !Task.isCancelled else {
+          throw CancellationError()
+        }
```

Comment:
I'm wondering if we actually want to check this: if the user handler has completed without throwing then -- presumably -- most of the hard work is done and we should just continue?

Reply:
Okay, let's leave it then.